notifier/service/service.go

```diff
@@ -0,0 +1,223 @@
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "notifier/service/Init").
+		Logger()
+	ctx = log.WithContext(ctx)
+
+	if opts.Webhook != nil && opts.AMQP != nil {
+		return nil, fmt.Errorf("notifier does not support both AMPQ and Webhook delivery")
+	}
+
+	// initialize store and dist lock pool
+	store, lockPool, err := storeInit(ctx, opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize store and lockpool: %v", err)
+	}
+
+	// kick off the poller
+	log.Info().Str("interval", opts.PollInterval.String()).Msg("intializing poller")
+	poller := notifier.NewPoller(opts.PollInterval, store, opts.Matcher)
+	c := poller.Poll(ctx)
+
+	// kick off the processors
+	log.Info().Int("count", processors).Msg("intializing processors")
+	for i := 0; i < processors; i++ {
+		// processors only use try locks
+		distLock := pgdl.NewLock(lockPool, 0)
+		p := notifier.NewProcessor(
+			i,
+			distLock,
+			opts.Indexer,
+			opts.Matcher,
+			store,
+		)
+		p.Process(ctx, c)
+	}
+
+	// kick off configured deliverer type
+	switch {
+	case opts.Webhook != nil:
+		if err := webhookDeliveries(ctx, opts, lockPool, store); err != nil {
+			return nil, err
+		}
+	case opts.AMQP != nil:
+		if err := amqpDeliveries(ctx, opts, lockPool, store); err != nil {
+			return nil, err
+		}
+	}
+
+	return &service{
+		store: store,
+	}, nil
+}
+
+func storeInit(ctx context.Context, opts Opts) (*postgres.Store, *sqlx.DB, error) {
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "notifier/service/storeInit").
+		Logger()
+	ctx = log.WithContext(ctx)
+
+	cfg, err := pgxpool.ParseConfig(opts.ConnString)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to parse ConnString: %v", err)
+	}
+	cfg.MaxConns = 30
+	pool, err := pgxpool.ConnectConfig(ctx, cfg)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create ConnPool: %v", err)
+	}
+
+	lockPool, err := sqlx.Connect("pgx", opts.ConnString)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create lock ConnPool: %v", err)
+	}
+
+	// do migrations if requested
+	if opts.Migrations {
+		log.Info().Msg("performing notifier migrations")
+		migrator := migrate.NewPostgresMigrator(lockPool.DB)
+		migrator.Table = migrations.MigrationTable
+		err := migrator.Exec(migrate.Up, migrations.Migrations...)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to perform migrations: %w", err)
+		}
+	}
+
+	log.Info().Msg("intializing notifier store")
+	store := postgres.NewStore(pool)
+	return store, lockPool, nil
+}
+
+func webhookDeliveries(ctx context.Context, opts Opts, lockPool *sqlx.DB, store notifier.Store) error {
+	log := zerolog.Ctx(ctx).With().
+		Str("component", "notifier/service/webhookInit").
+		Logger()
+	ctx = log.WithContext(ctx)
+	log.Info().Int("count", deliveries).Msg("intializing webhook deliverers")
+
+	ds := make([]*notifier.Delivery, 0, deliveries)
+	for i := 0; i < deliveries; i++ {
+		distLock := pgdl.NewLock(lockPool, 0)
+		wh, err := webhook.New(*opts.Webhook, nil)
```

Comment:
This has an embedded lock, it's not OK to copy.

Reply:
locks are gone.